### PR TITLE
Finetuning

### PR DESCRIPTION
**Pull request by jamesrequa**
Using VGG16 pre-trained on imagenet for fine tuning on the Planet Amazon dataset. Achieved 0.928 fbeta score after just 10 epochs (using 128x128 images).

### CODE REVIEW

File: notebooks/amazon_forest_notebook.py [hunk 1]
```diff
@@ -190,119 +191,121 @@
...
+from keras.callbacks import ModelCheckpoint, EarlyStopping, ReduceLROnPlateau
 
-train_losses, val_losses = [], []
-for learn_rate, epochs in zip(learn_rates, epochs_arr):
-    tmp_train_losses, tmp_val_losses, fbeta_score = classifier.train_model(learn_rate, epochs, batch_size, 
-                                                                           train_callbacks=[checkpoint, fbeta, csv])
-    train_losses += tmp_train_losses
-    val_losses += tmp_val_losses
+callbacks = [EarlyStopping(monitor='val_loss', patience=2, verbose=1, min_delta=1e-4),
+             ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=1, cooldown=0, verbose=1),
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: If we use `ReduceLROnPlateau` we don't need the `for` loop over our array of epochs/learning_rate right? I guess we can get rid of this thing :D

File: notebooks/amazon_forest_notebook.py [hunk 2]
```diff
@@ -190,119 +191,121 @@
...
+epochs_arr = [5, 3, 2]
+learn_rates = [1e-4, 1e-5, 1e-6]
+batch_size = 128
+train_generator = preprocessor.get_train_generator(batch_size)
+steps = len(X_train) // batch_size
 
-# ## Load Best Weights
+for learn_rate, epochs in zip(learn_rates, epochs_arr):
+    model.compile(optimizer=Adam(lr=learn_rate), loss='binary_crossentropy', metrics = ['accuracy'])
+    model.fit_generator(train_generator, steps, epochs, verbose=1,
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: I get an error here `Unable to open file: name = 'weights/weights.best.hdf5'`. I believe this is because you didn't commit the `weights` folder. In git if you try to commit an empty folder it will be ignored (I know it's kinda weird). You have to put 1 file in the folder at least. If you don't know what kind of "junk file" you want to put in the folder you can put either a `.DS_Store` or a `__init__.py` for folders containing python source files :)

File: notebooks/amazon_forest_notebook.py [hunk 3]
```diff
@@ -190,119 +191,121 @@
...
+# ## Make predictions
 
 # <codecell>
 
-fbeta_score
-
-# <markdowncell>
-
-# ## Make predictions
+def predict(batch_size=128):
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: Why did you move this method out of the `keras_helper.py` file? 🤔

File: notebooks/amazon_forest_notebook.py [hunk 4]
```diff
@@ -313,21 +316,31 @@
...
+# <markdowncell>
 
-# <codecell>
+# Now lets map our predictions to their tags by using the thresholds
 
-thresholds = [0.24, 0.2, 0.2, 0.2, 0.2, 0.14, 0.05, 0.2, 0.2, 0.25, 0.25, 0.24, 0.2, 0.25, 0.2, 0.2, 0.25]
+# <codecell>
 
-# <markdowncell>
+def map_predictions(predictions, thresholds):
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: Same here

File: src/custom_layers.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+class Scale(Layer):
+    '''Custom Layer for DenseNet used for BatchNormalization.
+
+    Learns a set of weights and biases used for scaling the input data.
+    the output consists simply in an element-wise multiplication of the input
+    and a sum of a set of constants:
+
+        out = in * gamma + beta,
+
+    where 'gamma' and 'beta' are the weights and biases larned.
```
Comment (godardt) [inline src/custom_layers.py]: Typo here on `larned` :)

File: src/custom_layers.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+from keras.engine import Layer, InputSpec
+try:
+    from keras import initializations
+except ImportError:
+    from keras import initializers as initializations
+import keras.backend as K
+
+class Scale(Layer):
```
Comment (godardt) [inline src/custom_layers.py]: Did you write this class by your own? If you didn't it's always a good idea to put the link where you found it in the document for later referencing :) . I'm a bit curious about why we need this class 🤔

File: src/data_helper.py [hunk 7]
```diff
@@ -105,7 +104,14 @@ def get_train_generator(self, batch_size):
...
 
                     # Convert to RGB and normalize
-                    img_array = np.asarray(img.convert("RGB"), dtype=np.float32) / 255
+                    img_array = np.asarray(img.convert("RGB"), dtype=np.float32)
+
+                    img_array = img_array[:, :, ::-1]
+                    # Zero-center by mean pixel
+                    img_array[:, :, 0] -= 103.939
+                    img_array[:, :, 1] -= 116.779
+                    img_array[:, :, 2] -= 123.68
```
Comment (godardt) [inline src/data_helper.py:113]: Oh nice! :octocat:  Does it means that we'll use the mean for **all** models from now? (Vgg, Resnet...)

File: src/data_helper.py [hunk 8]
```diff
@@ -181,7 +194,7 @@ def _get_validation_split(self):
         flatten = lambda l: [item for sublist in l for item in sublist]
         labels = list(set(flatten([l.split(' ') for l in train['tags'].values])))
         label_map = {l: i for i, l in enumerate(labels)}
-        inv_label_map = {i: l for l, i in label_map.items()}
+        #inv_label_map = {i: l for l, i in label_map.items()}
```
Comment (godardt) [inline src/data_helper.py]: We can remove this :)

File: src/data_helper.py [hunk 9]
```diff
@@ -144,10 +151,16 @@ def get_prediction_generator(self, batch_size):
...
-
                     # Convert to RGB and normalize
-                    img_array = np.asarray(img.convert("RGB"), dtype=np.float32) / 255
+                    img_array = np.asarray(img.convert("RGB"), dtype=np.float32)
+
+                    img_array = img_array[:, :, ::-1]
+                    # Zero-center by mean pixel
+                    img_array[:, :, 0] -= 103.939
+                    img_array[:, :, 1] -= 116.779
+                    img_array[:, :, 2] -= 123.68
```
Comment (godardt) [inline src/data_helper.py:161]: I really fucked it with this. You shouldn't have to modify the same code in 2 places :(

File: src/data_helper.py [hunk 10]
```diff
@@ -144,10 +151,16 @@ def get_prediction_generator(self, batch_size):
...
+                    img_array = img_array / 255
```
Comment (godardt) [inline src/data_helper.py:162]: Why now we divide by 255? 🤔

File: src/data_helper.py [hunk 11]
```diff
@@ -264,7 +285,14 @@ def _val_transform_to_matrices(self, *args):
         # Augment the image `img` here
 
         # Convert to RGB and normalize
-        img_array = np.array(img.convert("RGB"), dtype=np.float32) / 255
+        img_array = np.array(img.convert("RGB"), dtype=np.float32)
+        img_array = img_array[:, :, ::-1]
+        # Zero-center by mean pixel
+        img_array[:, :, 0] -= 103.939
+        img_array[:, :, 1] -= 116.779
+        img_array[:, :, 2] -= 123.68
```
Comment (godardt) [inline src/data_helper.py:292]: Uh.. 3 places. This code definitely needs a refactoring, sorry for that :(

File: src/data_helper.py [hunk 12]
```diff
@@ -264,7 +285,14 @@ def _val_transform_to_matrices(self, *args):
...
+        img_array = img_array / 255
```
Comment (godardt) [inline src/data_helper.py:293]: Again, why dividing by 255 for the test/val set and not with the train set? 🤔

File: src/densenet169.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+        x = conv_block(concat_feat, stage, branch, growth_rate, dropout_rate, weight_decay)
+        concat_feat = merge([concat_feat, x], mode='concat', concat_axis=concat_axis, name='concat_'+str(stage)+'_'+str(branch))
+
+        if grow_nb_filters:
+            nb_filter += growth_rate
+
+    return concat_feat, nb_filter
+
+
+def create_model():
```
Comment (godardt) [inline src/densenet169.py]: I don't see this method being called anywhere in the code. Are you sure we really need it? If we don't we should remove it as it will just add an "information junk burden" to the code :)

**Review by godardt — CHANGES_REQUESTED** [review]
Very good PR!! Good job! 🎉

File: notebooks/amazon_forest_notebook.py [hunk 14]
```diff
@@ -190,119 +191,121 @@
...
+from sklearn.metrics import fbeta_score
 
-# <codecell>
+def get_fbeta_score(model, X_valid, y_valid):
+    p_valid = model.predict(X_valid)
+    return fbeta_score(y_valid, np.array(p_valid) > 0.2, beta=2, average='samples')
 
-plt.plot(train_losses, label='Training loss')
-plt.plot(val_losses, label='Validation loss')
-plt.legend();
```
Comment (EKami) [inline notebooks/amazon_forest_notebook.py]: The plot, why did they disappear? 😱 🙀 
We can still pass the `History` callback to the training right?

Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: The plot, why did they disappear? 😱 🙀
We can still pass the History callback to the training right?

File: notebooks/amazon_forest_notebook.py (hunk 1 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: I see things such as `Epoch 00004: reducing learning rate to 4.999999873689376e-05.` So I believe we can definitely remove our for loop :D

File: notebooks/amazon_forest_notebook.py [hunk 15]
```diff
@@ -190,119 +191,121 @@
...
-print(fbeta.fbeta)
+X_train, y_train = preprocessor.X_train, preprocessor.y_train
+X_val, y_val = preprocessor.X_val, preprocessor.y_val
 
-# <markdowncell>
+epochs_arr = [5, 3, 2]
+learn_rates = [1e-4, 1e-5, 1e-6]
+batch_size = 128
+train_generator = preprocessor.get_train_generator(batch_size)
+steps = len(X_train) // batch_size
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: Using `//` isn't a good idea I think. I did that by the past (because intuitively it makes sense) and it resulted in Keras not training on the whole data. For whatever reason Keras figure out how to deal with `float` values for the steps. I know it's weird but it works 😛 .

**Review by godardt — CHANGES_REQUESTED** [review]

File: notebooks/amazon_forest_notebook.py [hunk 16]
```diff
@@ -190,119 +191,121 @@
...
+                   weights='imagenet',
+                   input_shape=img_dim)
 
-# <codecell>
+bn = BatchNormalization()(input_tensor)
+x = base_model(bn)
+x = Flatten()(x)
+output = Dense(17, activation='sigmoid')(x)
+
+model = Model(input_tensor, output)
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py]: We could move all this (line 218 to 227) to its own file as you did with `densenet169.py`. We can name it `vgg16.py` perhaps? :)

**Review by godardt — CHANGES_REQUESTED** [review]

File: notebooks/amazon_forest_notebook.py (hunk 3 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: Oh it's because you are applying it to the VGG model right? In this case see my comment [here](https://github.com/EKami/planet-amazon-deforestation/pull/14/files/9bee972a40beaa772e14b7595697ec163c1e49c9#r128918990), we should move this method to it's own file too :)

File: notebooks/amazon_forest_notebook.py (hunk 4 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: See my comment [here](https://github.com/EKami/planet-amazon-deforestation/pull/14/files/9bee972a40beaa772e14b7595697ec163c1e49c9#r128918990), we should move this method to it's own file too :)

File: notebooks/amazon_forest_notebook.py (hunk 1 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: I actually had originally thought it would still be nice to have a "learning rate schedule" on top of the `ReduceLROnPlateau`. Right now I'm retraining without the loop to see how `ReduceLROnPlateau` does on its own, but in this case I changed `factor` to 0.1 so that the learning rate decreases more (effectively creating our learning rate schedule on the fly).

File: notebooks/amazon_forest_notebook.py (hunk 2 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: Strange I actually did include the weights folder in the repo when I committed but I guess it didn't show up because the folder was empty like you say. So this time I'll make sure there is a file in there :)

**Comment by godardt** [comment]
https://github.com/EKami/planet-amazon-deforestation/pull/14#discussion_r128934730

@jamesrequa I also tried to use only LROnPlateau myself and it worked pretty well (got approximatively the same accuracy as the original code when running it on 20 epochs) but it's incompatible with `EarlyStopping` be careful. As the latter will stop the training if it sees that it plateau and ReduceLROnPlateau will try to reduce the learning rate.

File: notebooks/amazon_forest_notebook.py (hunk 3 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: This is because I'm not using the `AmazonKerasClassifier` class. The same goes for `map_predictions`. Only these two functions are being used from `keras_helper` file so I just moved them into the notebook. For this reason we could probably do without the `keras_helper` file entirely...Or do you have a better idea?

Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: For VGG model, its only like 8 lines of code (not counting the imports), do you think its really so much that it needs to be in a separate file and not just have it in the notebook? The reason DenseNet is its own separate file is because its very long and not a part of Keras so I couldn't just import it easily like I do with VGG.

Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: Sorry I didn't commit my message in time, please refer to my comment [here](https://github.com/EKami/planet-amazon-deforestation/pull/14#discussion_r128919042) :)

File: notebooks/amazon_forest_notebook.py (hunk 4 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: Yea this is because this function and `predict` are the only two functions I am using from `keras_helper`. I am not sure we should still keep that file just for two functions?

File: src/custom_layers.py (hunk 6 above)
Comment (jamesrequa) [reply src/custom_layers.py]: This class was created to help us convert DenseNet from caffe model to Keras. I am actually not using it in this PR so I can remove it and the DenseNet model as well.

File: notebooks/amazon_forest_notebook.py (hunk 3 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: Yes as you can also move the predict/map prediction methods to their own file :) . Creating a new file for a separate context is never a bad idea. Keeping multiples contexts into 1 file is :)

File: src/data_helper.py (hunk 7 above)
Comment (jamesrequa) [reply src/data_helper.py:113]: Yes this same preprocessing works for all imagenet pre-trained models which is basically every model we would want to use for finetuning :)

File: notebooks/amazon_forest_notebook.py (hunk 4 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: Moving them to a separate file like `vgg16.py` could be a good idea. Moving everything related to the VGG model to its own file is great I think :)

File: src/data_helper.py (hunk 10 above)
Comment (jamesrequa) [reply src/data_helper.py:162]: I have read that its still a good idea to have all image arrays as smaller numbers so this is why we divide by 255 after subtracting the mean.

File: src/data_helper.py (hunk 7 above)
Comment (godardt) [reply src/data_helper.py:113]: Nice!! :D But I think didn't get the whole story around why some ppl decides to normalize the values, other decides to remove the mean other do something else blah blah. I'm still a bit confused even if I understand that here we are forced to do that for finetunning.

File: src/data_helper.py (hunk 12 above)
Comment (jamesrequa) [reply src/data_helper.py:293]: We are dividing by 255 in the training set if you check the `ImageDataGenerator` we use `rescale=1./255`. So in this way actually training/validation/predicitons all consistently divide by 255.

File: src/densenet169.py (hunk 13 above)
Comment (jamesrequa) [reply src/densenet169.py]: Yes we don't use it because in the branch I was using DenseNet model but in the PR I switched to just use VGG and no longer using DenseNet. I will remove DenseNet model file and custom layer files.

File: src/data_helper.py (hunk 10 above)
Comment (godardt) [reply src/data_helper.py:162]: Oh ok. I still don't really get it lol. Why would smaller numbers would be quicker to process? I guess I'll figure this out with time ☹️

File: notebooks/amazon_forest_notebook.py (hunk 14 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: I honestly didn't find much use for it and because I was dealing with limited memory allocation I didn't want to waste any memory where it wasn't absolutely essential for training (i.e. so we can use a bigger batch size). If you disagree then I can add the history callback back in including the plot.

File: notebooks/amazon_forest_notebook.py (hunk 1 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: In order for `ReduceLROnPlateau` to be compatible with `EarlyStopping` we can actually use the `min_lr` argument to set a lower bound on the learning rate (i.e. `min_lr=1e-7`) so that `ReduceLROnPlateau` will stop reducing the learning rate beyond that.

File: notebooks/amazon_forest_notebook.py (hunk 4 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: OK I'll create a new file called `vgg16.py` which contains the model and also the two functions `predict` and `map_predictions`.

File: notebooks/amazon_forest_notebook.py (hunk 1 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: And actually they are technically compatible anyway because I have set `min_delta=1e-4` for `EarlyStopping` this means that unless `val_loss` improves by at least 1e-4 then its not considered any improvement meanwhile `ReduceLROnPlateau` will not reduce the lr as long as there is some small improvement regardless of how small it is.

File: notebooks/amazon_forest_notebook.py (hunk 14 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: Oh, I see. Actually I can augment your docker limit from 15gb to something like 20gb (it'll use the swap space tho). If I do that could we put back the curves? They were very useful to me 😢 . Deal?

File: notebooks/amazon_forest_notebook.py (hunk 1 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: Oh I see. On my side I only removed it without trying to make them both compatible with each other. My bad :(

File: notebooks/amazon_forest_notebook.py (hunk 14 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: Btw I don't think they take up too much space memory. It was only taking 2 `float32` per epochs (1 for the `train_loss`, the other for the `val_loss`). So lets say we ran it on 30 epochs = 30 * 4 = 120 bytes = 0.00012mb of memory space :D

**Comment by godardt** [comment]
You added `notebooks/weights/weights.best.hdf5` . This should be added to [`.gitignore`](https://github.com/EKami/planet-amazon-deforestation/blob/master/.gitignore) (so we don't commit large files (this one is 30mb)). We should only add a junk file to the `notebooks/weights/` folder such as `notebooks/weights/.DS_Store` :)

File: notebooks/amazon_forest_notebook.py (hunk 15 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: Sometimes I see it show `253 / 252` instead of `252 / 252` so this makes me wonder if its really seeing the full batches, do you think it would help to do `steps_per_epoch = len(X_train) / batch_size + 1` just to be safe?

File: notebooks/amazon_forest_notebook.py (hunk 14 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py]: OK no worries I am adding back the history callback in along with the plot :)

**Comment by jamesrequa** [comment]
@EKami All of the requested changes for this PR have now been completed. Let me know if all looks good for merging :)

File: notebooks/amazon_forest_notebook.py (hunk 15 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py]: I think yes. From my part when I'm not sure about that I always refer to the doc or try to find a response from an official repo. I just ran into this: https://github.com/fchollet/keras/pull/6301/files . And as you can see they use `/` instead of `//` on the Keras repo. I also had this issue myself when creating the batching PR, the last bits of the training set was "forgotten" while using `//` (which makes sense). 

I can't tell why but when I trained the model with `/` I've got better results than with `//`. Maybe because of the randomness of events? Maybe because we were dismissing some `primary` and `clear` images in the process? I can't tell.

File: .gitignore [hunk 17]
```diff
@@ -2,7 +2,8 @@
 input/*
 submission_file.csv
 notebooks/amazon_forest_notebook.ipynb
-notebooks/weights.best.hdf5
+notebooks/weights/*
+!notebooks/weights/.keep
```
Comment (godardt) [inline .gitignore:6]: Oh neat! I didn't know about this `!` 😄

File: src/keras_helper.py [hunk 18]
```diff
@@ -1,157 +0,0 @@
-import numpy as np
```
Comment (EKami) [inline src/keras_helper.py]: Bye bye little one 😢

File: notebooks/amazon_forest_notebook.py [hunk 19]
```diff
@@ -190,110 +191,78 @@
...
-# 
-# If you opted for a high `img_resize` then you may want to lower the `batch_size` to fit your images matrices into the GPU memory. With an `img_resize` of `(256, 256)` (the default size) and a batch_size of `64` you should at least have a GPU with 8GB or VRAM.
+# Here we define the model for finetuning
 
 # <codecell>
 
-batch_size = 64
-epochs_arr = [35, 15, 5]
-learn_rates = [0.002, 0.0002, 0.00002]
+model = vgg16.create_model(img_dim=(128, 128, 3))
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py:207]: This could be moved to `vgg16.py`, it's vgg related. Then you could create a method `get_summary()` to print the summary below :)

File: notebooks/amazon_forest_notebook.py [hunk 20]
```diff
@@ -190,110 +191,78 @@
...
-classifier = AmazonKerasClassifier(preprocessor)
-classifier.add_conv_layer()
-classifier.add_flatten_layer()
-classifier.add_ann_layer(len(preprocessor.y_map))
+history = History()
+callbacks = [history, 
+             EarlyStopping(monitor='val_loss', patience=3, verbose=1, min_delta=1e-4),
+             ReduceLROnPlateau(monitor='val_loss', factor=0.1, patience=1, cooldown=0, min_lr=1e-7, verbose=1),
+             ModelCheckpoint(filepath='weights/weights.best.hdf5', verbose=1, save_best_only=True, 
+                             save_weights_only=True, mode='auto')]
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py:222]: We can also move this to `vgg16.py`. Not sure tho. Your call :)

File: .gitignore [hunk 21]
```diff
@@ -2,7 +2,8 @@
 input/*
 submission_file.csv
 notebooks/amazon_forest_notebook.ipynb
-notebooks/weights.best.hdf5
```
Comment (godardt) [inline .gitignore:5]: You're still missing this one now. We need to put it back 😛

File: notebooks/amazon_forest_notebook.py [hunk 22]
```diff
@@ -190,110 +191,78 @@
...
 
-classifier.load_weights("weights.best.hdf5")
-print("Weights loaded")
+plt.plot(history.history['loss'])
+plt.plot(history.history['val_loss'])
+plt.title('model loss')
+plt.ylabel('loss')
+plt.xlabel('epoch')
+plt.legend(['train', 'validation'], loc='upper left')
+plt.show()
```
Comment (godardt) [inline notebooks/amazon_forest_notebook.py:247]: Very neat! Thanks :D

**Review by godardt — CHANGES_REQUESTED** [review]
This PR is awesome! Just 2-3 little things to change and we're good 🎉

**Comment by godardt** [comment]
amazon_forest_notebook
There is something not right here. Is it?

**Comment by godardt** [comment]
amazon_forest_notebook
Btw the training stopped earlier than expected just after the `ReduceLROnPlateau` has decreased the learning rate. But I believe it's expected?

File: .gitignore (hunk 17 above)
Comment (jamesrequa) [reply .gitignore:6]: Yep this makes it so any file in weights folder will not get loaded

File: notebooks/amazon_forest_notebook.py (hunk 19 above)
Comment (jamesrequa) [reply notebooks/amazon_forest_notebook.py:207]: We are still currently using keras in the notebook for the model functions like `compile` and `fit_generator`, so unless we moved literally everything in vgg16.py I think its OK to just leave it as is. Do you think we should move everything to the helper file? This makes everything a lot more high level in the notebook but also hides a lot of the inner details of the network functions (making it more challenging for the user to tune the network within the notebook).

**Comment by jamesrequa** [comment]
@EKami 
For your first question, on the plot yes this is to be expected as the network quickly drops to a low loss and gets less and less decrease in loss over subsequent epochs. If we were to stop training at "convergence" which is much earlier in the epoch cycle we would end up with a much lower fbeta score. So in this case it makes sense to let it continue training according to our schedule set by `EarlyStopping` and`ReduceLROnPlateau`.

For your second question, yes the training has stopped where it should be after `ReduceLROnPlateau` has decreased to its minumum learning rate value (`min_lr=1e-7`).

File: notebooks/amazon_forest_notebook.py (hunk 19 above)
Comment (godardt) [reply notebooks/amazon_forest_notebook.py:207]: Well ideally the notebook should only be used to present results, not to tune anything but I didn't respect this rule from the very start so I don't know. Put (or leave) this line where you think is best :)

**Comment by godardt** [comment]
@jamesrequa Oh I see, thanks for these explanations :)

**Review by godardt — APPROVED** [review]
